feat: add Using Rust crates in Elixir blog post (#1975)
mainmatter/mainmatter.com: src/posts/2023-01-30-using-rust-crates-in-elixir.md

small private project. I searched for some packages and the most notable one for
Elixir is [elixir-pdf](https://github.com/andrewtimberlake/elixir-pdf), but it
only offers to create PDFs without additional manipulation. That was fine by me
since I wanted an excuse to do some more things in Rust anyway, and I found
[lopdf](https://github.com/J-F-Liu/lopdf).

In the beginning, I had written my program as a regular project in Rust i.e.
`cargo new my_pdf` and then it was happy coding 🙂. I wanted the program to take
some configuration and create/edit a PDF file based on it.

After I played around with it, I started wondering “How much work would it be to
make it an Elixir NIF?”.

Turns out - there’s not a whole lot of additional stuff we need to do in order
to comfortably use crates inside Elixir projects.

[Rustler](https://github.com/rusterlium/rustler) is a library for writing
[Erlang NIFs](https://www.erlang.org/doc/tutorial/nif.html) in a very easy and
straightforward way.

There are about 2 things we need to keep in mind and take care of:

- Error handling: there needs to be something that maps Rust errors into Elixir
  atoms so that we could easily handle them.

> - Rustler traits: typically your functions and structs will need to derive from

Reviewer: ```suggestion
- Rustler traits: typically your functions and structs will need to implement
```

I still thing this is the wrong wording in this context. One way to implement a trait is to **derive the implementation** but you don't **derive from a trait** as far as I understand it.

  appropriate traits so they can be usable inside Elixir (only the public parts
  of it of course).

## Setup

Setting up a project with Rustler is fairly easy. You can find a link to
instructions here:
[https://github.com/rusterlium/rustler#getting-started](https://github.com/rusterlium/rustler#getting-started)

I’ll assume you have both Elixir and Rust installed in your development
environment.

1. Create a new Elixir project `mix new rustler_pdf`

> 2. Add `:rustler` dependency in `mix.exs`

Reviewer: ```suggestion
2. Add the `:rustler` dependency in `mix.exs`
```

   ```elixir
    # mix.exs
     defp deps do
       [
         {:rustler, "~> 0.26.0"}
       ]
     end
   ```

> 3. Download packages `mix deps.get`

Reviewer: ```suggestion
3. Download packages with `mix deps.get`
```

or

```suggestion
3. Download packages: `mix deps.get`
```

4. Setup Rustler
   1. Run `mix rustler.new`

>    2. On “Module name” prompt type in `RustlerPdf` (Name of your Elixir module

Reviewer: ```suggestion
   2. On the “Module name” prompt, type in `RustlerPdf` (Name of your Elixir module
```

      that Rustler registers NIFs to)

>    3. On “Library name” prompt type in `rustlerpdf` (Name of your cargo crate)

Reviewer: ```suggestion
   3. On the “Library name” prompt, type in `rustlerpdf` (Name of your cargo crate)
```

5. Configure Rustler

>    1. Add Rustler behaviour in `rustler_pdf.ex`

Reviewer: ```suggestion
   1. Add the Rustler behaviour in `rustler_pdf.ex`
```

   end
   ```

   2. Make sure Rust’s boilerplate is this:

   ```rust
   // native/src/lib.rs

   #[rustler::nif]
   fn add(a: i64, b: i64) -> i64 {
       a + b
   }

   rustler::init!("Elixir.RustlerPdf", [add]);
   ```

6. Test that the setup works
   1. Run `iex -S mix`
   2. Execute `RustlerPdf.add(1,2)`
   3. See if it successfully outputs `3`

## Implementing the PDF program

I’ll mostly showcase contracts and interfaces of the Rust part of the program.
If you're interested in the full implementation, you can find it
[here](https://github.com/BobrImperator/rustler_pdf).

First off let’s take a look at pure Rust structs and functions.

```rust
// Enum that determinates value formatting
pub enum FieldType {
    Money,
    Text,
    Slotted,
}

// Struct describing text
pub struct PdfWriterOperation {
    page_number: i32,
    font: (String, i32),
    dimensions: (f64, f64),
    field_type: FieldType,
    value: Option<String>,
}

// Struct which is the program's input
pub struct PdfWriterConfiguration {
    input_file_path: Option<String>,
    output_file_path: String,
    operations: Vec<PdfWriterOperation>,
}

> pub fn read_config() -> PdfWriterConfiguration {

Reviewer: I still think it's confusing that this is called `read_config` when it seems to return a static example PDF configuration?

Author: That's a nuance, what it returns is important and not how it's called exactly.

Reviewer: Sure, but it's confusing for readers when the name of the function has no relation to what it's doing. When you read this you ask yourself what you missed.

    PdfWriterConfiguration {
        input_file_path: None,
        output_file_path: "PIT-8C-modified.pdf".to_string(),
        operations: vec![
            PdfWriterOperation {
                page_number: 0,
                font: ("F1".to_string(), 10),
                dimensions: (462.82, 55.92),
                value: Some("120.99".to_string()),
                field_type: FieldType::Money,
            },
            PdfWriterOperation {
                page_number: 0,
                font: ("F1".to_string(), 10),
                dimensions: (43.32, 347.81),
                value: Some("41.0".to_string()),
                field_type: FieldType::Money,
            },
        ],
    }
}

pub fn create_pdf(config: PdfWriterConfiguration) -> Result<(), std::io::Error> {}
```

This is the overview of pretty much the entirety of the Rust implementation
“minus” the `lopdf` interaction. In general, the idea is that given
`PdfWriterConfiguration` a `create_pdf` function will create a PDF file. Then I
want to use those methods directly in Elixir.

> The only relevant part of implementation here are the structs themselves, as you

Reviewer: ```suggestion
The only relevant part of the implementation here are the structs themselves, as you
```

> can see they have `number`, `string`, `tuple`, `struct` and `enum` Rust types.

Reviewer: ```suggestion
can see they have number, `string`, `tuple`, `struct` and `enum` Rust types.
```

I think it's confusing to format number as code here as that suggest it's actually a type while you're really using it to summarize i32 et al.

`NifTaggedEnum` in order to use such Enums:

```rust
use rustler::NifTaggedEnum;

#[derive(NifTaggedEnum)]
pub enum FieldType {
    Money(String),
    Text(String),
    Slotted(String),
}`
```

Then in Elixir they are represented as a tuple like this:

```elixir
{:slotted, "some value"}
```

### Back to Rustlerizing - Functions

```rust
#[rustler::nif]
pub fn read_config() -> PdfWriterConfiguration {
  priv_read_config()
}

fn priv_read_config() -> PdfWriterConfiguration {
  PdfWriterConfiguration {
        input_file_path: None,
        output_file_path: "PIT-8C-modified.pdf".to_string(),
        operations: vec![
            PdfWriterOperation {
                page_number: 0,
                font: ("F1".to_string(), 10),
                dimensions: (462.82, 55.92),
                value: Some("120.99".to_string()),
                field_type: FieldType::Money,
            },
            PdfWriterOperation {
                page_number: 0,
                font: ("F1".to_string(), 10),
                dimensions: (43.32, 347.81),
                value: Some("41.0".to_string()),
                field_type: FieldType::Money,
            },
        ],
    }
}
```

> `read_config` function is simple enough that it doesn’t require any additional

Reviewer: ```suggestion
The `read_config` function is simple enough that it doesn’t require any additional
```

treatment. Since it’s only delivering hardcoded data and is not dealing with any

> kind of IO, it just needs an attribute
> 
> `#[rustler::nif]`.

Reviewer: ```suggestion
kind of IO, it just needs the `#[rustler::nif]` attribute.
```

I’m also splitting the functions into the plain Rust functions and the public

> ones that Rustler exposes to Elixir. This most definitely should be built as

Reviewer: ```suggestion
ones that Rustler exposes to Elixir. This most definitely should be built as a
```

regular Rust crate and another Rustler ‘bridge’ module. I decided to just add
prefixes here - for the purpose of this post, I believe this simplifies things a
bit.

```rust
use rustler::{Atom, Env, Error as RustlerError, NifStruct, NifUnitEnum, Term};
use std::io::Error as IoError;
use std::io::ErrorKind as IoErrorKind;

> // Creates atoms module using rustler macro

Reviewer: ```suggestion
// Creates an atoms module using the rustler macro
```

mod atoms {
    rustler::atoms! {
        ok,
        error,
        eof,

        // Posix
        enoent, // File does not exist
        eacces, // Permission denied
        epipe, // Broken pipe
        eexist, // File exists

        unknown // Other error
    }
}

// Translates std library errors into Rustler atoms
fn io_error_to_term(err: &IoError) -> Atom {
    match err.kind() {
        IoErrorKind::NotFound => atoms::enoent(),
        IoErrorKind::PermissionDenied => atoms::eacces(),
        IoErrorKind::BrokenPipe => atoms::epipe(),
        IoErrorKind::AlreadyExists => atoms::eexist(),
        _ => atoms::unknown(),
    }
}

#[rustler::nif]
pub fn create_pdf(env: Env, config: PdfWriterConfiguration) -> Result<Term, RustlerError> {
    match priv_create_pdf(config) {
        Ok(()) => Ok(atoms::ok().to_term(env)),
        Err(ref error) => return Err(RustlerError::Term(Box::new(io_error_to_term(error)))),
    }
}

fn priv_create_pdf(config: PdfWriterConfiguration) -> Result<(), std::io::Error> {}
```

> `create_pdf` is slightly more complicated as it involves some error handling.

Reviewer: ```suggestion
`create_pdf` is slightly more complicated than `read_config` as it involves some error handling.
```

I think that's what you mean?

`priv_create_pdf` is dealing with creating an actual file on the filesystem so
it might result in an error.

> In order to handle such situation, we’ve defined a module called `atoms` which

Reviewer: ```suggestion
In order to handle that situation, we’ve defined a module called `atoms` which
```

uses the `rustler::atoms!` macro to create methods to simplify creating and
decoding atoms.

`io_error_to_term` takes an `std::io::error`, matches on the kind of error and
translates that into an `Atom`.

> As you’ve probably noticed `create_pdf` expects an `env` argument which is

Reviewer: ```suggestion
As you’ve probably noticed, `create_pdf` expects an `env` argument which is
```

always present with all NIFs. Env is used by Rustler for communication and

> encoding/decoding data between BEAM.

Reviewer: ```suggestion
encoding/decoding data between the BEAM and Rust.
```

_"between"_ requires two things :)

`read_config` and `add` in the previous examples are not using it so it’s
possible to omit its declaration.

### Exporting NIFs

```rust
rustler::init!("Elixir.RustlerPdf", [read_config, create_pdf]);
```

> At the end of the file change the `rustler::init!` macro to the above: this lets

Reviewer: ```suggestion
At the end of the file, change the `rustler::init!` macro to the above: this lets
```

This is an important one – without the comma, it's kind of hard to know what you mean

> Rustler know to bind `read_config` function to `Elixir.RustlerPdf` module.

Reviewer: ```suggestion
Rustler know to bind `read_config` function to the `Elixir.RustlerPdf` module.
```

## The Elixir part

Once the Rust crate exports the functions, we need to let the Elixir side know
what functions are expected to be bound to the module. Note that knowing the

> functions arity is important here, otherwise you’ll get errors saying that a

Reviewer: ```suggestion
functions' arities is important here, otherwise you’ll get errors saying that a
```

    )
  end
```

```elixir
Operating System: macOS
CPU Information: Intel(R) Core(TM) i7-9750H CPU @ 2.60GHz
Number of Available Cores: 12
Available memory: 16 GB
Elixir 1.14.2
Erlang 23.3.4.17

Benchmark suite executing with the following configuration:
warmup: 2 s
time: 10 s
memory time: 1 s
reduction time: 0 ns
parallel: 1
inputs: none specified
Estimated total run time: 26 s

Benchmarking create_pdf ...
Benchmarking e_create_pdf ...

Name                   ips        average  deviation         median         99th %
create_pdf          2.02 K        0.49 ms    ±56.09%        0.48 ms        0.70 ms
e_create_pdf        0.96 K        1.04 ms    ±42.20%        1.03 ms        1.22 ms

Comparison:
create_pdf          2.02 K
e_create_pdf        0.96 K - 2.11x slower +0.55 ms

Memory usage statistics:

Name            Memory usage
create_pdf           1.65 KB
e_create_pdf         1.17 KB - 0.71x memory usage -0.47656 KB
```

### Conclusion

I've shown how you can make use of Rustler in order to add a functionality to
your Elixir program which natively might not exist. I hope I also managed to
make you consider using Rust for your next high-peformant and type-safe Elixir
module.

> In my opinion Rust and Elixir are a great match. Rust offers amazing processing

Reviewer: ```suggestion
In my opinion, Rust and Elixir are a great match. Rust offers amazing processing
```
